### PR TITLE
MAT-425 MVP MAT - Add Convert to FHIR button to grid toolbar - REAL Services

### PR DESCRIPTION
**Pull request by ghost**
https://jira.cms.gov/browse/MAT-425

As a measure developer, I want to be able to take an action in the MAT so that I can convert a QDM measure to FHIR.

In this story, we will perform a measure conversion from the  Convert to FHIR toolbar button to both MAT measure grids.

The mockup for the UI is found here:  https://balsamiq.cloud/sg54b2t/pt8orhx/rDBB8

Acceptance Criteria:

0.0 This is behind the FhirConvV1 feature flag.

1.0 When the Convert to FHIR button is clicked from one of the grid toolbars in theMeasure Library page the conversion services are executed.

2.0 If user selects a QDM (CQL) measure that has already been converted and answers 'Yes' to the prompt to continue, the conversion services are executed.

3.0 When the conversion is done, the validation report will be generated and displayed in a new browser page.

4.0 After the conversion, the FHIR measure is displayed in the grid.

5.0 If user selects a QDM (CQL) measure that has already been converted and the associated FHIR measure still exists display an 'Are you sure' prompt

     5.1 Issue the message as a dialog box. Prompt text: "Are you sure you want to convert this measure again? The existing FHIR measure will be overwritten."  Provide a "Yes" button and a "No" button

    5.2 If the user clicks Yes, we'll close the dialog box.    In a future story, when the user answers Yes, we will proceed with the conversion as outlined in MAT-114 The existing (previously converted) FHIR measure will be overwritten.  Note: this will be done in MAT-352

    5.3 If the user clicks No, close the dialog box and do nothing.

6.0 The prior validations from MAT-352 mock services story still work.

### CODE REVIEW

File: src/main/java/mat/dao/service/DAOService.java [hunk 1]
```diff
@@ -1,498 +0,0 @@
-package mat.dao.service;
-
-import org.springframework.beans.factory.annotation.Autowired;
```
Comment (serhii-ilin) [inline src/main/java/mat/dao/service/DAOService.java]: It was not used

File: src/main/java/mat/server/Application.java [hunk 2]
```diff
@@ -82,20 +88,10 @@ public LocalSessionFactoryBean sessionFactory(@Autowired DataSource dataSource,
...
-        hibernateProperties.setProperty("hibernate.show_sql", "false");
-        hibernateProperties.setProperty("hibernate.default_batch_fetch_size", "20");
-        hibernateProperties.setProperty("hibernate.jdbc.batch_size", "50");
-        hibernateProperties.setProperty("hibernate.order_updates", "true");
-        hibernateProperties.setProperty("hibernate.connection.release_mode", "auto");
-        hibernateProperties.setProperty("hibernate.cache.use_second_level_cache", "true");
-        hibernateProperties.setProperty("hibernate.dialect", "org.hibernate.dialect.MySQL5Dialect");
-        hibernateProperties.setProperty("entityInterceptor", "mat.dao.impl.AuditInterceptor");
-        return hibernateProperties;
+    @ConfigurationProperties(prefix = "spring.jpa.properties")
```
Comment (serhii-ilin) [inline src/main/java/mat/server/Application.java]: gradually moving towards spring-boot

File: src/main/java/mat/server/MatSessionListener.java [hunk 3]
```diff
@@ -1,50 +0,0 @@
...
-	
-	/** The Constant logger. */
-	private static final Log logger = LogFactory.getLog(MatSessionListener.class);
-
-	/* (non-Javadoc)
-	 * @see javax.servlet.http.HttpSessionListener#sessionCreated(javax.servlet.http.HttpSessionEvent)
-	 */
-	@Override
-	public void sessionCreated(HttpSessionEvent arg0) {
-		logger.debug("Session created: " + arg0.getSession().getId());
```
Comment (serhii-ilin) [inline src/main/java/mat/server/MatSessionListener.java]: It has no business logic or any value. It was just printing a stack trace on session created/destroyed

File: src/main/java/mat/server/service/FhirValidationReportService.java [hunk 4]
```diff
@@ -70,18 +71,16 @@ public String getFhirConversionReportForMeasure(String measureId) throws IOExcep
...
-        URL path = FhirValidationReportService.class.getResource("report.json");
-        logger.info("Calling FHIR conversion validation service for measure: "+ measureId);
-        //TODO: Replace following line with actual call to FHIR conversion service once it is available on AWS
-        // and delete report.json mock data
-        return new ObjectMapper()
-                .readValue(new File(path.getFile()),
-                        ConversionResultDto.class);
+        logger.info("Calling FHIR conversion validation service for measure: " + measureId);
+        try {
+            return fhirOrchestrationGatewayService.validate(measureId, isDraft);
```
Comment (serhii-ilin) [inline src/main/java/mat/server/service/FhirValidationReportService.java]: Invoke real service

File: src/main/java/mat/server/service/FhirValidationReportService.java [hunk 5]
```diff
@@ -98,22 +97,17 @@ private String generateValidationReport(ConversionResultDto conversionResultDto,
...
-                    getCqlConversionErrors();
-
-            //ValueSet FHIR validation errors
-            List<ValueSetValidationResult> valueSetFhirValidationErrors = conversionResultDto.
-                    getValueSetConversionResults().
-                    getValueSetFhirValidationErrors();
+                    getMeasureConversionResults().getMeasureFhirValidationResults();
+
+            // Library FHIR validation errors
+            List<LibraryConversionResults> libraryConversionResults = conversionResultDto.getLibraryConversionResults();
```
Comment (serhii-ilin) [inline src/main/java/mat/server/service/FhirValidationReportService.java]: It's a temporary fix to make the code compilable and do integration testing. The logic should be adjusted during real validation service implementation.

File: src/main/java/mat/server/LoggedInUserUtil.java [hunk 6]
```diff
@@ -2,178 +2,131 @@
...
+    /**
+     * Gets the token.
+     *
+     * @return the token
+     */
+    private static UsernamePasswordAuthenticationToken getToken() {
+        //re-factored to support Anonymous user US 439
+        UsernamePasswordAuthenticationToken token = null;
+        if (SecurityContextHolder.getContext() != null) {
+            if (SecurityContextHolder.getContext().getAuthentication() != null) {
```
Comment (RohitKandimalla) [inline src/main/java/mat/server/LoggedInUserUtil.java]: Can't we use single if statement ?

Comment (serhii-ilin) [reply src/main/java/mat/server/LoggedInUserUtil.java]: I didn't change this piece of code :) It was just re-formatted. Tabs->Spaces

Comment (serhii-ilin) [reply src/main/java/mat/server/LoggedInUserUtil.java]: and removed some redundant obsolete comments

If you check "Hide whitespace changes" you will see cleaner PR view

Comment (RohitKandimalla) [reply src/main/java/mat/server/LoggedInUserUtil.java]: ok

**Review by RohitKandimalla — APPROVED** [review]

File: src/main/java/mat/server/service/impl/FhirMeasureServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    public FhirMeasureServiceImpl(FhirOrchestrationGatewayService fhirOrchestrationGatewayService, MeasureLibraryService measureLibraryService, MeasureCloningService measureCloningService, MeasureDAO measureDAO, PlatformTransactionManager txManager) {
+        this.fhirOrchestrationGatewayService = fhirOrchestrationGatewayService;
+        this.measureLibraryService = measureLibraryService;
+        this.measureCloningService = measureCloningService;
+        this.measureDAO = measureDAO;
+        this.transactionTemplate = new TransactionTemplate(txManager);
+    }
+
+    @Override
+    public FhirConvertResultResponse convert(ManageMeasureSearchModel.Result sourceMeasure, String loggedinUserId) throws MatException {
```
Comment (adongare) [inline src/main/java/mat/server/service/impl/FhirMeasureServiceImpl.java]: Can we have some documentation for this method. It seems to be the main method responsible for conversion process

**Review by adongare — APPROVED** [review]

File: src/main/java/mat/server/service/impl/FhirMeasureServiceImpl.java (hunk 7 above)
Comment (serhii-ilin) [reply src/main/java/mat/server/service/impl/FhirMeasureServiceImpl.java]: I think we should prepare a sequence diagram for the conversion

**Pull request merged**